### PR TITLE
amd_media - update for GPU mode

### PR DESCRIPTION
**Pull request by rrawther**
fixed parameter string parsing error for encoder.
fixed gpu buffer sharing for encoder.

### CODE REVIEW

File: amd_openvx_extensions/amd_media/decoder.cpp [hunk 1]
```diff
@@ -583,7 +585,9 @@ vx_status CLoomIoMediaDecoder::Initialize()
     }
 
     for (int i = 0; i < DECODE_BUFFER_POOL_SIZE; i++) {
-        err = hipHostMalloc((void **)&decodeBuffer[i], gpuOffset + gpuStride * height);
+        int buf_height = height;
+        if (outputFormat == AV_PIX_FMT_NV12)  buf_height = height + (height>>1);
+        err = hipHostMalloc((void **)&decodeBuffer[i], gpuOffset + gpuStride * buf_height);
```
Comment (asalmanp) [inline amd_openvx_extensions/amd_media/decoder.cpp:590]: where is the decodeBuffer used?

**Review by asalmanp — CHANGES_REQUESTED** [review]

File: amd_openvx_extensions/amd_media/encoder.cpp [hunk 2]
```diff
@@ -191,34 +195,14 @@ CLoomIoMediaEncoder::~CLoomIoMediaEncoder()
...
-                mem[i] = nullptr;
-            } else {
-                status = hipFree(mem[i]);
-                if (status != hipSuccess) {
-                    vxAddLogEntry(NULL, VX_FAILURE, "ERROR: hipFree(%p) failed (%d)\n", mem[i], status);
-                }
-            }
-        }
+    if (m_enableUserBufferGPU) {
+    #if ENABLE_OPENCL
```
Comment (asalmanp) [inline amd_openvx_extensions/amd_media/encoder.cpp:199]: Where do you free the mem (gpu memory)?

**Review by asalmanp — CHANGES_REQUESTED** [review]

File: amd_openvx_extensions/amd_media/decoder.cpp [hunk 3]
```diff
@@ -567,7 +567,9 @@ vx_status CLoomIoMediaDecoder::Initialize()
     #endif
         ERROR_CHECK_NULLPTR(cmdq);
         for (int i = 0; i < DECODE_BUFFER_POOL_SIZE; i++) {
-            mem[i] = clCreateBuffer(context, CL_MEM_READ_WRITE, gpuOffset + gpuStride * height, nullptr, nullptr);
+            int buf_height = height;
+            if (outputFormat == AV_PIX_FMT_NV12)  buf_height = height + (height>>1);
+            mem[i] = clCreateBuffer(context, CL_MEM_READ_WRITE, gpuOffset + gpuStride * buf_height, nullptr, nullptr);
```
Comment (asalmanp) [inline amd_openvx_extensions/amd_media/decoder.cpp:572]: Where do you free the mem[i] (gpu memory)?

**Review by asalmanp — CHANGES_REQUESTED** [review]

File: amd_openvx_extensions/amd_media/encoder.cpp [hunk 4]
```diff
@@ -331,57 +338,32 @@ vx_status CLoomIoMediaEncoder::Initialize()
...
-                return VX_FAILURE;
-            }
-        } else {
-            err = hipMalloc(&mem[i], offset + stride * height);
-            if (err != hipSuccess) {
-                vxAddLogEntry(NULL, VX_FAILURE, "ERROR: hipMalloc(%p) => %d (failed)\n", mem[i], err);
-                return VX_FAILURE;
-            }
+        int bufHeight = (inputFormat == AV_PIX_FMT_NV12) ? (height + (height>>1)) : height;
+        hostBuffer = (uint8_t *)malloc(offset + stride * bufHeight);
```
Comment (asalmanp) [inline amd_openvx_extensions/amd_media/encoder.cpp:360]: please change to hipHostMalloc to create pinned host memory.

File: amd_openvx_extensions/amd_media/encoder.cpp [hunk 5]
```diff
@@ -191,34 +195,14 @@ CLoomIoMediaEncoder::~CLoomIoMediaEncoder()
...
-                if (status != hipSuccess) {
-                    vxAddLogEntry(NULL, VX_FAILURE, "ERROR: hipFree(%p) failed (%d)\n", mem[i], status);
-                }
-            }
-        }
+    if (m_enableUserBufferGPU) {
+    #if ENABLE_OPENCL
+        if (cmdq) clReleaseCommandQueue(cmdq);
+    #elif ENABLE_HIP
+        if (hostBuffer) free(hostBuffer);
```
Comment (asalmanp) [inline amd_openvx_extensions/amd_media/encoder.cpp:202]: please change to hipHostFree

**Review by asalmanp — CHANGES_REQUESTED** [review]

File: amd_openvx_extensions/amd_media/encoder.cpp (hunk 4 above)
Comment (rrawther) [reply amd_openvx_extensions/amd_media/encoder.cpp]: done

File: amd_openvx_extensions/amd_media/encoder.cpp (hunk 5 above)
Comment (rrawther) [reply amd_openvx_extensions/amd_media/encoder.cpp]: done

**Comment by rrawther** [comment]
fixed review comments.

**Review by asalmanp — APPROVED** [review]

**Pull request merged**